src/mono/mono/metadata/class-init.c

```diff
@@ -4197,65 +4197,76 @@ mono_class_set_runtime_vtable (MonoClass *klass, MonoVTable *vtable)
+build_variance_search_table_inner (MonoClass *klass, MonoVarianceSearchTableEntry *buf, int buf_size, int *buf_count, MonoClass *current) {
+	if (!m_class_is_interfaces_inited (current)) {
 		ERROR_DECL (error);
-		mono_class_setup_interfaces (klass, error);
+		mono_class_setup_interfaces (current, error);
 		return_if_nok (error);
 	}
-	guint c = m_class_get_interface_count (klass);
+
+	guint c = m_class_get_interface_count (current);
 	if (c) {
-		MonoClass **ifaces = m_class_get_interfaces (klass);
+		MonoClass **ifaces = m_class_get_interfaces (current);
 		for (guint i = 0; i < c; i++) {
 			MonoClass *iface = ifaces [i];
 			// Avoid adding duplicates or recursing into them.
 			if (index_of_class (iface, buf, *buf_count) >= 0)
 				continue;
 
 			if (mono_class_has_variant_generic_params (iface)) {
-				g_assert (*buf_count < buf_size);
-				buf[*buf_count] = iface;
+				int index = *buf_count;
 				(*buf_count) += 1;
+				g_assert (index < buf_size);
+				buf[index].klass = iface;
+				buf[index].depth = current->idepth;
+				buf[index].offset = mono_class_interface_offset (klass, iface);
 			}
 
-			build_variance_search_table_inner (iface, buf, buf_size, buf_count);
+			build_variance_search_table_inner (klass, buf, buf_size, buf_count, iface);
 		}
 	}
+
+	if (current->parent)
+		build_variance_search_table_inner (klass, buf, buf_size, buf_count, current->parent);
 }
 
+typedef struct VarianceSearchTable {
+	int count;
+	MonoVarianceSearchTableEntry entries[1]; // a total of count items, at least 1
+} VarianceSearchTable;
+
 // Only call this with the loader lock held
 static void
 build_variance_search_table (MonoClass *klass) {
-	// FIXME: Is there a way to deterministically compute the right capacity?
 	int buf_size = m_class_get_interface_offsets_count (klass), buf_count = 0;
-	MonoClass **buf = g_alloca (buf_size * sizeof(MonoClass *));
-	MonoClass **result = NULL;
+	MonoVarianceSearchTableEntry *buf = g_alloca (buf_size * sizeof(MonoVarianceSearchTableEntry));
+	VarianceSearchTable *result = NULL;
 	memset (buf, 0, buf_size * sizeof(MonoClass *));
-	build_variance_search_table_inner (klass, buf, buf_size, &buf_count);
+	build_variance_search_table_inner (klass, buf, buf_size, &buf_count, klass);
 
 	if (buf_count) {
-		guint bytes = buf_count * sizeof(MonoClass *);
-		result = mono_mem_manager_alloc (m_class_get_mem_manager (klass), bytes);
-		memcpy (result, buf, bytes);
+		guint bytes = ((buf_count - 1) * sizeof(MonoVarianceSearchTableEntry)) + sizeof(VarianceSearchTable);
+		result = (VarianceSearchTable *)mono_mem_manager_alloc (m_class_get_mem_manager (klass), bytes);
+		result->count = buf_count;
+		memcpy (result->entries, buf, bytes);
 	}
-	klass->variant_search_table_length = buf_count;
 	klass->variant_search_table = result;
 	// Ensure we do not set the inited flag until we've stored the result pointer
 	mono_memory_barrier ();
 	klass->variant_search_table_inited = TRUE;
```

Comment:
Should we save a bit?  you could use `klass->variant_search_table = (void*)(intptr_t)-1;` to signal that it's initialized but not useful